pkg/controller/mssql/grant/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane-contrib/provider-sql/apis/mssql/v1alpha1"
+	"github.com/crossplane-contrib/provider-sql/pkg/clients/mssql"
+	"github.com/crossplane-contrib/provider-sql/pkg/clients/xsql"
+)
+
+const (
+	errTrackPCUsage = "cannot track ProviderConfig usage"
+	errGetPC        = "cannot get ProviderConfig"
+	errNoSecretRef  = "ProviderConfig does not reference a credentials Secret"
+	errGetSecret    = "cannot get credentials Secret"
+
+	errNotGrant        = "managed resource is not a Grant custom resource"
+	errGrant           = "cannot grant"
+	errRevoke          = "cannot revoke"
+	errCannotGetGrants = "cannot get current grants"
+
+	maxConcurrency = 5
+)
+
+// Setup adds a controller that reconciles Grant managed resources.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.GrantGroupKind)
+
+	t := resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{})
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1alpha1.GrantGroupVersionKind),
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), usage: t, newDB: mssql.New}),
+		managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithPollInterval(10*time.Minute),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Grant{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrency,
+		}).
+		Complete(r)
+}
+
+type connector struct {
+	kube  client.Client
+	usage resource.Tracker
+	newDB func(creds map[string][]byte, database string) xsql.DB
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return nil, errors.New(errNotGrant)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	// ProviderConfigReference could theoretically be nil, but in practice the
+	// DefaultProviderConfig initializer will set it before we get here.
+	pc := &v1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	// We don't need to check the credentials source because we currently only
+	// support one source (MSSQLConnectionSecret), which is required and
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
```

Comment:
This is quite a standard boilerplate that we have in all providers. I agree that this could still be refactored but considering we don't expect frequent changes here, I prefer to keep it as is for now.